pageserver/page_api/src/client.rs

```diff
@@ -0,0 +1,234 @@
+use std::convert::TryInto;
+use std::pin::Pin;
+use std::io::Error;
+use std::io::ErrorKind;
+
+use bytes::Bytes;
+use futures_core::Stream;
+use tonic::transport::Channel;
+use tonic::{Request, Status, Streaming};
+use tonic::metadata::AsciiMetadataValue;
+use tonic::metadata::errors::InvalidMetadataValue;
+use futures::StreamExt;
+
+use utils::id::TenantId;
+use utils::id::TimelineId;
+use utils::shard::ShardIndex;
+
+use crate::proto;
+use crate::model;
+
+#[derive(Clone)]
+pub struct Client {
+    client: proto::PageServiceClient<Channel>,
+    tenant_id: String,
+    timeline_id: String,
+    shard_id: String,
+    auth_header: Option<String>,
+}
+
+fn convert_metadata_err(e: InvalidMetadataValue, s: String) -> Error {
+    let io_err = Error::new(
+        ErrorKind::InvalidInput,
+        format!("{} header was invalid: {}", s, e),
+    );
+    io_err
+}
+
+impl Client {
+
+    pub async fn new(connstring: String,
+                    tenant_id: TenantId,
+                    timeline_id: TimelineId,
+                    shard_id: ShardIndex,
+                    auth_header: Option<String>) -> Result<Self, Error> {
+
+        let mut client = proto::PageServiceClient::connect(connstring).await
+            .map_err(|e| {
+                Error::new(ErrorKind::Other, format!("Failed to connect to page service: {}", e))
+            })?;
+
+        // Attempt to parse headers, so that parsing fails early. Later, we can assume
+        // that the headers are valid ASCII strings.
+        let _tenant_ascii : AsciiMetadataValue = tenant_id.to_string().try_into()
+            .map_err(|e: InvalidMetadataValue| {
+                convert_metadata_err(InvalidMetadataValue::from(e), "tenant-id".to_string())
+            })?;
+
+        let _timeline_ascii : AsciiMetadataValue = timeline_id.to_string().try_into()
+            .map_err(|e: InvalidMetadataValue| {
+                convert_metadata_err(InvalidMetadataValue::from(e), "timeline-id".to_string())
+            })?;
+
+        let _shard_ascii : AsciiMetadataValue = shard_id.clone().to_string().try_into()
+            .map_err(|e: InvalidMetadataValue| {
+                convert_metadata_err(InvalidMetadataValue::from(e), "shard-id".to_string())
+            })?;
```

Comment:
Instead of re-parsing these for every request, let's just store the parsed value in `self`.